### PR TITLE
Define {build,host,target}Platform in stdenv

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

See https://github.com/NixOS/nixpkgs/issues/27069 In particular, make derivation needs to be aware of the target platforms sometimes, and right now it can't be. An alternative would be to make stdenv unaware, but tack on `mkDerivation` separately so it alone is aware.

This depends on https://github.com/NixOS/nixpkgs/pull/27045 to be merged first.

###### Things done

No native hashes are changed. Some cross hashes are changed, but only changed because `stdenv.is*` properly evaluates to the host platform when cross compiling too.

[I had PR that was supposed to fix that `stdenv.is*` behavior before, but it only did so natively when `hostPlatform == buildPlatform`. This partial fix was necessary because `meta.platform` and it were coupled, and `meta.platform` still needs to be checked against the build platform until I can audit the packages we cross compile today. Down the road I'll fix those packages, but also teach meta checking that there's 3 platforms, not 1.]

Built the manual locally (along with the other Travis-built derivations for Linux).

---

CC @vcunat

### CODE REVIEW

**Comment by vcunat** [comment]
Accidental nitpick: the author name in 2314b481e shows as some insane character sequence to me.

**Comment by Ericson2314** [comment]
@vcunat oh hahah, I saw in `gitk` and assumed it was just a rendering issue. Fixed now.

**Comment by vcunat** [comment]
I can't get [one XKCD](http://bobby-tables.com/) out of my head :-)

**Comment by edolstra** [comment]
Going from ~2.8s to ~3.5s is not a small slowdown!

**Comment by Ericson2314** [comment]
Ok, with the doc changes I made, this should be ready. I'm going to hop on the subway, but be very tempted to merge when I got off :D. Last call, reviewers!

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            lib.concatMap (input: input.__propagatedSandboxProfile or []) (lib.concatLists (lib.reverseList propagatedDependencies'));
+          computedImpureHostDeps =
+            lib.unique (lib.concatMap (input: input.__propagatedImpureHostDeps or []) (stdenv.extraBuildInputs ++ lib.concatLists (lib.reverseList dependencies')));
+          computedPropagatedImpureHostDeps =
+            lib.unique (lib.concatMap (input: input.__propagatedImpureHostDeps or []) (lib.concatLists (lib.reverseList propagatedDependencies')));
+        in
+        {
+          builder = attrs.realBuilder or stdenv.shell;
+          args = attrs.args or ["-e" (attrs.builder or ./default-builder.sh)];
+          inherit stdenv;
```
Comment (Et7f3) [inline pkgs/stdenv/generic/make-derivation.nix:80]: This create infinite recursion we can't compute expect.all or ocaml.all

Comment (Et7f3) [inline pkgs/stdenv/generic/make-derivation.nix:80]: This create infinite recursion we can't compute expect.all or ocam.all

Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:80]: This is probably not the commit you wished to comment on?

Comment (Et7f3) [reply pkgs/stdenv/generic/make-derivation.nix:80]: I tested until that part but yes it might be before because it is a code movement.

**Pull request merged**